### ISSUE TITLE
Non-determinism when already fixing random seeds

### ISSUE THREAD

**Issue by KelestZ** [rosinality/stylegan2-pytorch#78]
Hi, I noticed the model has non-determinism even I fixed the random seed. For example, when I trained it on a rendered chair dataset. The same noise has different image outputs.
Pair 1: 
![image](https://user-images.githubusercontent.com/27111467/82744323-931d7000-9d3c-11ea-9c81-0f04d29df698.png)
![image](https://user-images.githubusercontent.com/27111467/82744325-96b0f700-9d3c-11ea-88ed-262ad0357cf8.png)

Pair 2:
![image](https://user-images.githubusercontent.com/27111467/82744366-f3141680-9d3c-11ea-8f3f-b1ccea63ce68.png)
![image](https://user-images.githubusercontent.com/27111467/82744367-f5767080-9d3c-11ea-9bf0-c65f178ef0d7.png)

The issue also exists on some natural image datasets, but not obvious as it on this chair dataset.
Why could this happen and is there a way to fix it?

**Comment by rosinality** [comment]
Some cuda operations has nondeterministic behavior, so you will need to set like this:

```python
torch.backends.cudnn.deterministic = True
torch.backends.cudnn.benchmark = False
```

**Comment by KelestZ** [comment]
It doesn't work. So my steps are like this: z --> w=G_mapping(z) -->I=G_syn(w), and I repeat w=G_mapping(z) -->I=G_syn(w) with the same z for multiple times in one run. It gave me different output Imgs.
I already fixed torch.manual_seed and the cuddn benchmark. I guess there is still some other random parts?

**Comment by rosinality** [comment]
Random noises will not much affect the results. Other sources that could cause nondeterminism is random indexes that change of latents happen, but this will not be the problem unless you have used two latent vectors.

I thought your problem was about training nondeterminism, but it seems like that this not the case. Could you track the layers that divergence of two results happen?

**Comment by KelestZ** [comment]
This happens during inference time. I've used torch.eq() to test the two latent vectors and they were the same. I agreed that random noise doesn't have this issue on natural images as I tested.
If all these factors are not causing the issue, I was doubting if stylegan2 can fit this kind of  3D rendered dataset...

**Comment by rosinality** [comment]
Do you mean that identity changes only when `randomize_noise=True`? Hmm I haven't seen cases like that. But I think you can also train model without these additional noises.

**Comment by thoppe** [comment]
@KelestZ I noticed this as well and (I think) track down the source of confusion. Turning off random noise isn't enough due to the following code:

``` python
class NoiseInjection(nn.Module):
    def __init__(self):
        super().__init__()

        self.weight = nn.Parameter(torch.zeros(1))

    def forward(self, image, noise=None):
        if noise is None:
            
            batch, _, height, width = image.shape
            noise = image.new_empty(batch, 1, height, width).normal_()
           

        return image + self.weight * noise
```

So if `noise==None` as expected when turning the noise off, a new matrix of noise is still generated! The fix for this is to generate the noise yourself and pass it into the noise parameter in the forward pass for the generator:

``` python
# To reproduce, need to save the noise matrix each time
noise = [getattr(model.noises, f"noise_{i}") for i in range(model.num_layers)]

img0, _ = clf([z0,], noise=noise, input_is_latent=False, randomize_noise=False)
img1, _ = clf([z0,], noise=noise, input_is_latent=False, randomize_noise=False)
```

Without fixing the noise, you'll get (slightly) different results each time.
